pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitSlicedRangeIndexReader.java

```diff
@@ -0,0 +1,150 @@
+  private final long _offset;
+  private final long _min;
+  private volatile SoftReference<RangeBitmap> _rangeBitmapRef;
+
+  public BitSlicedRangeIndexReader(PinotDataBuffer dataBuffer) {
+    _dataBuffer = dataBuffer;
+    long offset = 0;
+
+    int version = dataBuffer.getInt(offset);
+    assert version == BitSlicedRangeIndexCreator.VERSION : "invalid version";
+    offset += Integer.BYTES;
+
+    _min = dataBuffer.getLong(offset);
+    offset += Long.BYTES;
+
+    this._offset = offset;
+  }
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getMatchingDocIds(int min, int max) {
+    return queryRangeBitmap(Math.max(min, _min) - _min, max - _min);
+  }
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getMatchingDocIds(long min, long max) {
+    return queryRangeBitmap(Math.max(min, _min) - _min, max - _min);
+  }
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getMatchingDocIds(float min, float max) {
+    return queryRangeBitmap(FPOrdering.ordinalOf(min), FPOrdering.ordinalOf(max));
+  }
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getMatchingDocIds(double min, double max) {
+    return queryRangeBitmap(FPOrdering.ordinalOf(min), FPOrdering.ordinalOf(max));
+  }
+
+  // this index supports exact matches, so always return null for partial matches
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getPartiallyMatchingDocIds(int min, int max) {
+    return null;
+  }
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getPartiallyMatchingDocIds(long min, long max) {
+    return null;
+  }
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getPartiallyMatchingDocIds(float min, float max) {
+    return null;
+  }
+
+  @Nullable
+  @Override
+  public ImmutableRoaringBitmap getPartiallyMatchingDocIds(double min, double max) {
+    return null;
+  }
+
+  private ImmutableRoaringBitmap queryRangeBitmap(long min, long max) {
+    RangeBitmap rangeBitmap = getRangeBitmap();
+    RoaringBitmap lte = rangeBitmap.lte(max);
+    RoaringBitmap gte = rangeBitmap.gte(min);
+    lte.and(gte);
+    return lte.toMutableRoaringBitmap();
+  }
+
+  private RangeBitmap getRangeBitmap() {
+    SoftReference<RangeBitmap> ref = _rangeBitmapRef;
+    RangeBitmap rangeBitmap;
+    if (ref == null) {
+      rangeBitmap = mapRangeBitmap();
+      UPDATER.compareAndSet(this, null, new SoftReference<>(rangeBitmap));
+    } else {
+      rangeBitmap = ref.get();
+      if (rangeBitmap == null) {
+        // it's been cleared so reinitialize
+        rangeBitmap = mapRangeBitmap();
+        UPDATER.compareAndSet(this, ref, new SoftReference<>(rangeBitmap));
+      }
+    }
+    return rangeBitmap;
+  }
+
+  private RangeBitmap mapRangeBitmap() {
+    // note that this is a very cheap operation, no deserialization is required
+    ByteBuffer buffer = _dataBuffer.toDirectByteBuffer(_offset, (int) (_dataBuffer.size() - _offset));
+    return RangeBitmap.map(buffer);
+  }
```

Comment:
@npawar I imagine you would like this changed so there's no caching here?

Reply:
Removed by https://github.com/apache/pinot/pull/7454/commits/04c1cfbb1fa6433ec41a96f991c66253d06dfa1c

Reply:
Yes, thank you! This will help with bringing in the acquire-release semantics